employee-rostering-server/src/main/java/org/optaweb/employeerostering/server/tenant/TenantRestServiceImpl.java

```diff
@@ -60,6 +68,36 @@ public Tenant addTenant(RosterState intialRosterState) {
         return intialRosterState.getTenant();
     }
 
+    @Override
+    @Transactional
+    public Boolean removeTenant(Integer tenantId) {
+        // Dependency order: Shift, EmployeeAvailability, ShiftTemplate,
+        // Employee, Spot, Skill,
+        // RosterParametrization, RosterState
+        Tenant tenant = getTenant(tenantId);
+        if (tenant == null) {
+            throw new IllegalArgumentException("There is no tenant with id (" + tenantId + ").");
+        }
+        deleteAllEntitiesBelongingToTenantOfType(tenantId, Shift.class, "Shift.findAll");
+        deleteAllEntitiesBelongingToTenantOfType(tenantId, EmployeeAvailability.class, "EmployeeAvailability.findAll");
+        deleteAllEntitiesBelongingToTenantOfType(tenantId, ShiftTemplate.class, "ShiftTemplate.findAll");
+        deleteAllEntitiesBelongingToTenantOfType(tenantId, Employee.class, "Employee.findAll");
+        deleteAllEntitiesBelongingToTenantOfType(tenantId, Spot.class, "Spot.findAll");
+        deleteAllEntitiesBelongingToTenantOfType(tenantId, Skill.class, "Skill.findAll");
+        deleteAllEntitiesBelongingToTenantOfType(tenantId, RosterParametrization.class, "RosterParametrization.find");
+        deleteAllEntitiesBelongingToTenantOfType(tenantId, RosterState.class, "RosterState.find");
+        entityManager.remove(tenant);
+        return true;
+    }
+
+    private <T extends AbstractPersistable> void deleteAllEntitiesBelongingToTenantOfType(Integer tenantId, Class<T> type, String queryName) {
+        List<T> entitiesBelongingToTenant = entityManager.createNamedQuery(queryName, type).setParameter("tenantId", tenantId)
+                                                         .getResultList();
+        for (T entity : entitiesBelongingToTenant) {
+            entityManager.remove(entity);
```

Comment:
This is slow.

Add a JPA-QL per entity type that does "delete where tenantId = $tenantId"

Reply:
Maybe consider using Criteria API? https://stackoverflow.com/questions/23846618/bulk-delete-in-jpa-2-1-using-the-criteria-api